Add item requests and gateway

Eugenpak opened the pull request:
Проект ShareIt. Техническое задание. Часть 3
add-item-requests-and-gateway

Eugenpak committed:
Добавить запрос вещи пакет request.

Eugenpak committed:
Сохранить изменения в пакете booking.

Eugenpak committed:
Сохранить изменения в пакете item.

Eugenpak committed:
Сохранить изменения в пакете user (UserServiceImpl).

Eugenpak committed:
Сохранить изменения в пакете main.resources

Eugenpak committed:
Добавить тесты для пакета booking.

Eugenpak committed:
Добавить тесты для пакета user.

Eugenpak committed:
Добавить тесты для пакета item.

Eugenpak committed:
Добавить тесты для пакета request.

Eugenpak committed:
Добавить тесты для пакета exception.

Eugenpak committed:
Сохранить ShareItTests

Eugenpak committed:
Изменить RestExceptionHandlerTest

Eugenpak committed:
Удалить RestExceptionHandlerTest

Eugenpak committed:
Создать server (перенос src).

Eugenpak committed:
Создать Dockerfile и pom.xml для модуля server.

Eugenpak committed:
Добавить main class ShareItServer для модуля server.

Eugenpak committed:
Изменить application.properties для модуля server.

Eugenpak committed:
Создать gateway (добавить шаблон src).

Eugenpak committed:
Добавить ShareItGateway для модуля gateway.

Eugenpak committed:
Создать Dockerfile и pom.xml для модуля gateway.

Eugenpak committed:
Добавить дополнительные данные.

Eugenpak committed:
Добавить docker-compose.yml

Eugenpak committed:
Изменить родительский pom.xml

Eugenpak committed:
Сохранить пакет booking в модуле gateway.

Eugenpak committed:
Сохранить пакет item в модуле gateway.

Eugenpak committed:
Сохранить пакет user в модуле gateway.

Eugenpak committed:
Сохранить пакет request в модуле gateway.

Eugenpak committed:
Сохранить доп. в модуле gateway.

Eugenpak committed:
Сохранить application.properties в модуле gateway.

Eugenpak committed:
Сохранить Dockerfile в модуле gateway.

Eugenpak committed:
Сохранить Dockerfile в модуле server.

Eugenpak committed:
Checkstyle Spr-16-v-01

Eugenpak committed:
Checkstyle Spr-16-v-02

Eugenpak committed:
Checkstyle Spr-16-v-03

Eugenpak committed:
Изменить Dockerfile в модуле server.

Eugenpak committed:
Изменить пакет booking в модуле gateway.

Eugenpak committed:
Сохранить пакет validation в модуле gateway.

Eugenpak committed:
Изменить пакет exception в модуле gateway.

Eugenpak committed:
Сохранить тесты в модуле gateway.

Eugenpak committed:
Изменить Dockerfile и application.properties в модуле gateway.

Eugenpak committed:
Очистить код в тестах в модуле server.

Eugenpak committed:
Очистить код в пакете booking в модуле server.

Eugenpak committed:
Очистить код в пакете item в модуле server.

Eugenpak committed:
Очистить код в пакете request в модуле server.

Eugenpak committed:
Очистить код в пакете user в модуле server.

Eugenpak committed:
Удалить Marker.java в модуле server.

Eugenpak committed:
Изменить application.properties в модуле server.

Eugenpak committed:
Checkstyle Spr-16-v-04

avfyodorov commented on gateway/src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -0,0 +1,67 @@
+	private final BookingClient bookingClient;
+
+	@GetMapping
+	public ResponseEntity<Object> getBookings(@RequestHeader("X-Sharer-User-Id") long userId,
```
Не помню, давал ссылку или нет. На всякий случай повторю. 🙂

`ResponseEntity<?> ........`

Изучаем `ResponseEntity<?>` и избавляемся от него в контроллерах Spring

Если будет время и желание, посмотрите, пожалуйста, вот эту статью. 
https://habr.com/ru/articles/675716/

Eugenpak replied:
Здравствуйте, Александр! Замечание принял на заметку. Статью посмотрел. Проработаю статью позднее. Спасибо , что обратили внимание на эту тему.

avfyodorov commented on gateway/src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -0,0 +1,67 @@
+	}
+
+	@PostMapping
+	public ResponseEntity<Object> bookItem(@RequestHeader("X-Sharer-User-Id") long userId,
```
long userId,

Здесь и дальше.
К этим параметрам можно было бы добавить аннотацию Positive.  
Конечно, на Ваше усмотрение.

Eugenpak replied:
Исправил.

avfyodorov commented on gateway/src/main/java/ru/practicum/shareit/exception/RestExceptionHandler.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @ExceptionHandler
+    public ResponseEntity<ErrorResponse> handleThrowable(final Throwable e) {
```
Throwable --> Exception

Тут правильнее указать Exception, а не Throwable..... Error и его наследников не принято ловить.
.------------------
Все исключения делятся на 4 вида, которые на самом деле являются классами, унаследованными друг от друга.

Класс Throwable
Самым базовым классом для всех исключений является класс Throwable. В классе Throwable содержится код, который записывает текущий стек-трейс вызовов функций в массив. 

В оператор throw можно передать только объект класса-наследника Throwable. И хотя теоретически можно написать код вида throw new Throwable(); — так обычно никто не делает. Главная цель существования класса Throwable — единый класс-предок для всех исключений.

Класс Error
Следующим классом исключений является класс Error — прямой наследник класса Throwable. Объекты типа Error (и его классов-наследников) создает Java-машина в случае каких-то серьезных проблем. Например, сбой в работе, нехватка памяти, и т.д.

Обычно вы как программист ничего не можете сделать в ситуации, когда в программе возникла ошибка типа Error: слишком серьезна такая ошибка. Все, что вы можете сделать — уведомить пользователя, что программа аварийно завершается или записать всю известную информацию об ошибке в лог программы.

Класс Exception
Исключения типа Exception (и RuntimeException) — это обычные ошибки, которые возникают во время работы многих методов. Цель каждого выброшенного исключения — быть захваченным тем блоком catch, который знает, что нужно сделать в этой ситуации.

avfyodorov commented on gateway/src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -0,0 +1,60 @@
+
+    @GetMapping
+    public ResponseEntity<Object> getAll(@RequestHeader("X-Sharer-User-Id") Long userId,
+                                         @RequestParam(value = "from", required = false, defaultValue = "0") @Min(0) Integer from,
```
defaultValue уже подразумевает, что параметра может и не быть. Поэтому вполне можно не писать  required = false

Eugenpak replied:
Исправил

avfyodorov commented on gateway/src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -0,0 +1,60 @@
+                                            @RequestParam(name = "text") String text,
+                                            @RequestParam(value = "from", required = false, defaultValue = "0") @Min(0) Integer from,
+                                            @RequestParam(value = "size", required = false, defaultValue = "10") @Min(1) Integer size) {
+        return itemClient.getAllByText(userId, text, from, size);
```
Лучше было бы прямо здесь проверить входной параметр text. И, если текст запроса пустой, или недопустимый, например, состоит из пробелов, то прямо отсюда и вернуть результат, не обращаться к серверу. Вернуть можно, например, пустую коллекцию.

avfyodorov reviewed (commented):
Добрый вечер, Евгений! 

Хорошая работа, от меня всего пара уточнений.

Eugenpak committed:
Исправить замечания:
https://github.com/Eugenpak/java-shareit/pull/3#discussion_r2127254171
https://github.com/Eugenpak/java-shareit/pull/3#discussion_r2127252301
https://github.com/Eugenpak/java-shareit/pull/3#discussion_r2127250520
https://github.com/Eugenpak/java-shareit/pull/3#discussion_r2127246387

Eugenpak committed:
Checkstyle Spr-16-v-05

avfyodorov reviewed (commented):
Добрый день, Евгений!

Замечаний нет.
Работа принята.
